### PR TITLE
Implement the hackage and nixpkgs meta analyzers

### PR DESCRIPTION
**Pull request by MangoIV**
### Description

This solves part 2 of #3545 

### Addressed Issue

#3545 

### Additional Details

this depends on #3546 

the apache http client utils version 4 don't implement brotli decompression; I redid some of the common http client code to keep this scoped; a future refactoring could remove the extraneous code. 

### Checklist

- [x] I have read and understand the [contributing guidelines](../CONTRIBUTING.md#pull-requests)
- [x] This PR implements an enhancement, and I have provided tests to verify that it works as intended
- [ ] This PR introduces new or alters existing behavior, and I have updated the [documentation](https://github.com/DependencyTrack/dependency-track/tree/master/docs/_docs) accordingly

### CODE REVIEW

File: src/main/java/org/dependencytrack/tasks/repositories/NixpkgsMetaAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        this.baseUrl = DEFAULT_CHANNEL_URL;
+        HashMap<String, String> newLatestVersion = new HashMap<>();
+
+        try (final CloseableHttpResponse packagesResponse = processHttpRequest5()) {
+            if (packagesResponse != null && packagesResponse.getCode() == HttpStatus.SC_OK) {
+                final var entity = packagesResponse.getEntity();
+                if (entity != null) {
+                    // TODO(mangoiv): is this the fastest way we can do this?
+                    final var entityString = EntityUtils.toString(new DecompressingEntity(entity, new BrotliInputStreamFactory()));
+                    final var packages = new JSONObject(entityString).getJSONObject("packages").toMap().values();
```
Comment (MangoIV) [inline src/main/java/org/dependencytrack/tasks/repositories/NixpkgsMetaAnalyzer.java:41]: this line errors with `NoClassDefFoundError org/brotli/dec/BrotliInputStream`

**Comment by MangoIV** [comment]
what's missing: 
- [ ] invoke the refreshing job every now and then
- [ ] make Brotli decompression not error 
- [ ] the usual chores.

**Comment by MangoIV** [comment]
so if my theory is true, then this is some truly cursed shit: 
- at runtime, the http client library looks for presence of the brotli dec class 
- if it doesn't exist, it doesn't try to decode brotli encoded files but still removes the decoding header and goes on 
- if it does exist, it actually converts it into a decompressing entity and then goes on decoding it

**Comment by MangoIV** [comment]
Hi, I would like some input on where to best invoke the update of the nixpkgs meta analyzer singleton. Thanks in advance!

**Comment by emil-wire** [comment]
@nscuro I don't want to pester you and understand you have a lot on your mind. We'd be grateful if you could provide some guidance at your earliest convenience to make this PR a success and land this feature in DT. Thank you!

**Comment by nscuro** [comment]
@MangoIV @emil-wire Hey, will have a look at this ASAP and provide feedback. Thank you for your contributions! :)

**Comment by nscuro** [comment]
@MangoIV Instead of using a singleton of `NixpkgsMetaAnalyzer`, and updating it every now and then, you could use a read-through cache with "expire-after-write" eviction. The [Caffeine](https://github.com/ben-manes/caffeine) cache library is already available in DT.

So the first time a task tries to access nix versions, it will fetch and parse `packages.json`, and put it in the cache. Other tasks trying to access it are blocked until the cache population finishes. After that, they can be quickly served from said cache. The cache will expire after a given interval (e.g. 1h), and the whole thing repeats.

The benefit of this is that you won't have to deal with concurrency controls such as locking on your own. Also refreshing happens implicitly and as such doesn't require a dedicated scheduled task. 

Because we'd want to refresh the entire package index at once, the entire index would be a single cache entry:

```java
class NixPkgMetaAnalyzer {

    private static final Cache<String, Map<String, String>> CACHE = Caffeine.newBuilder()
            .expireAfterWrite(60, TimeUnit.MINUTES)
            .maximumSize(1)
            .build();

    MetaModel analyze(Component component) {
        Map<String, String> latestVersions = CACHE.get("nixpkgs", cacheKey -> {
            final var versions = new HashMap<String, String>();
            // TODO: Download and parse package index
            return versions;
        });

        // TODO: Do things with latestVersions
    }
```

More details here: https://github.com/ben-manes/caffeine/wiki/Population#manual

**Comment by MangoIV** [comment]
Thank you for the answer, that looks very nice! 

It looks like a fairly large hammer for the tiny nail in trying to get into the wall but if it’s already used across the codebase, this looks like a feasible option; I’ll adjust accordingly.

**Comment by MangoIV** [comment]
https://github.com/package-url/packageurl-java/commit/73b2f510a3ad09461ee52ada13e67a9ebd804dbf

btw; this got merged, i.e. theoretically I could do this minor cleanup; however upstream didn't do a release yet so idk if this is worth the effort.

**Comment by MangoIV** [comment]
tests pass locally, let's see if CI is merciful today :pleading_face:

**Comment by MangoIV** [comment]
@nscuro is this how you'd imagine it? what can I do to make this PR progress?

**Review by nscuro — APPROVED** [review]
Thanks @MangoIV!

As a future enhancement, we'd want to improve the parsing of the nixpkgs catalogue. At the moment, the entire thing is deserialized at once, which causes quite a large amount of memory allocations:

image

A more efficient way would be to leverage a streaming API such as the one provided by Jackson. That allows us to only deserialize fields we actually care about. We make use of that when parsing CVE data from the NVD already:

https://github.com/DependencyTrack/dependency-track/blob/3e8f93fb06465cea80723392c4e761ef94e20b8b/src/main/java/org/dependencytrack/parser/nvd/NvdParser.java#L82-L107

**Comment by MangoIV** [comment]
It’s a tradeoff. I would be reluctant to make a guess on whether it’s wise to use streaming here without thorough benchmarks.

**Pull request merged**